goyacc/format_yacc.go

```diff
@@ -0,0 +1,581 @@
+		}
+	}
+	return nil
+}
+
+func joinNames(names NameArr) string {
+	var sb strings.Builder
+	for _, name := range names {
+		sb.WriteString(" ")
+		sb.WriteString(getTokenComment(name.Token, spanStringLayout))
+		sb.WriteString(name.Token.Val)
+	}
+	return sb.String()
+}
+
+func printSingleName(f format.Formatter, name *parser.Name, maxCharLength int) error {
+	if hasComments(name) {
+		_, err := f.Format("\n%s\n", strings.Join(name.Token.Comments, "\n"))
+		if err != nil {
+			return err
+		}
+	}
+	if name.LiteralStringOpt != nil && name.LiteralStringOpt.Token != nil {
+		strLit := fmt.Sprintf(" %s", name.LiteralStringOpt.Token.Val)
+		_, err := f.Format("%-*s%s\n", maxCharLength, name.Token.Val, strLit)
+		return err
+	} else {
+		_, err := f.Format("%s\n", name.Token.Val)
+		return err
+	}
+}
+
+type NameArr []*parser.Name
+
+func (ns NameArr) span(pred func(*parser.Name) bool) (NameArr, NameArr) {
+	first := ns.takeWhile(pred)
+	second := ns[len(first):]
+	return first, second
+}
+
+func (ns NameArr) takeWhile(pred func(*parser.Name) bool) NameArr {
+	for i, def := range ns {
+		if pred(def) {
+			continue
+		}
+		return ns[:i]
+	}
+	return ns
+}
+
+func (ns NameArr) findMaxLength() int {
+	maxLen := -1
+	for _, s := range ns {
+		if len(s.Token.Val) > maxLen {
+			maxLen = len(s.Token.Val)
+		}
+	}
+	return maxLen
+}
+
+func hasComments(n *parser.Name) bool {
+	return len(n.Token.Comments) != 0
+}
+
+func noComment(n *parser.Name) bool {
+	return !hasComments(n)
+}
+
+func containsActionInRule(rule *parser.Rule) bool {
+	for _, b := range rule.Body {
+		if _, ok := b.(*parser.Action); ok {
+			return true
+		}
+	}
+	return false
+}
+
+type RuleArr []*parser.Rule
+
+func printRules(f format.Formatter, rules RuleArr) (err error) {
+	var lastRuleName string
+	for _, rule := range rules {
+		if rule.Name.Val == lastRuleName {
+			cmt := getTokenComment(rule.Token, divStringLayout)
+			_, err = f.Format("\n%s|\t%i", cmt)
+		} else {
+			cmt := getTokenComment(rule.Name, divStringLayout)
+			_, err = f.Format("\n\n%s%s:%i\n", cmt, rule.Name.Val)
+		}
+		if err != nil {
+			return err
+		}
+		lastRuleName = rule.Name.Val
+
+		if err = printRuleBody(f, rule); err != nil {
+			return err
+		}
+		if _, err = f.Format("%u"); err != nil {
+			return err
+		}
+	}
+	_, err = f.Format("\n%%%%\n")
+	return err
+}
+
+type ruleItemType int8
+
+const (
+	identRuleItemType      ruleItemType = 1
+	actionRuleItemType     ruleItemType = 2
+	strLiteralRuleItemType ruleItemType = 3
+)
+
+func printRuleBody(f format.Formatter, rule *parser.Rule) error {
+	firstRuleItem, counter := rule.RuleItemList, 0
+	for ri := rule.RuleItemList; ri != nil; ri = ri.RuleItemList {
+		switch ruleItemType(ri.Case) {
+		case identRuleItemType, strLiteralRuleItemType:
+			term := fmt.Sprintf(" %s", ri.Token.Val)
+			if ri == firstRuleItem {
+				term = term[1:]
+			}
+			cmt := getTokenComment(ri.Token, divStringLayout)
+
+			if _, err := f.Format(escapePercent(cmt)); err != nil {
```

Comment:
I think `f.Format("%s", cmt)` should be enough. If we use `/* ... */` instead of `// ...` it should handle multi-line comments fine? (ditto for everything involving `escapePercent()`)

Reply:
`cmt` contains characters like `\n`, which is used by `format.Formatter` (when `%i` is enabled) to determine whether it should print an indention or not.

To use the `%i`/`%u` mechanism, the string contains `\n` must be passed into `Format` as the first argument. So it's different between
```
f.Format("%s", cmt)
```
and 
```
f.Format(cmt)
```

Reply:
@tangenta I understand the difference between `Format(s)` vs `Format("%s", s)`, what I'm saying is that perhaps we could avoid `Format(s)`.